java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java

```diff
@@ -446,28 +393,32 @@ private Resolution findMethod(Env env, JavaType callSite, JavaType site, String
     // look in site members
     for (JavaSymbol symbol : site.getSymbol().members().lookup(name)) {
       if (symbol.kind == JavaSymbol.MTH) {
-        JavaSymbol best = selectBest(env, callSite, argTypes, symbol, bestSoFar.symbol, autoboxing);
+        JavaSymbol best = selectBest(env, callSite, argTypes, typeParams, symbol, bestSoFar.symbol, autoboxing);
         if(best == symbol) {
           bestSoFar = Resolution.resolution(best);
-          bestSoFar.type = resolveTypeSubstitution(((JavaType.MethodJavaType) best.type).resultType, site);
-          JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) best;
-          bestSoFar.type = handleTypeArguments(typeParams, bestSoFar.type, methodSymbol);
+          if (!isConstructor(symbol)) {
+            bestSoFar.type = typeSubstitutionSolver.getReturnType((JavaSymbol.MethodJavaSymbol) best, site, typeParams, argTypes);
+          }
         }
       }
     }
     //look in supertypes for more specialized method (overloading).
     if (superclass != null) {
       Resolution method = findMethod(env, callSite, superclass, name, argTypes, typeParams);
-      method.type = resolveTypeSubstitution(resolveTypeSubstitution(method.type, superclass), site);
-      JavaSymbol best = selectBest(env, callSite, argTypes, method.symbol, bestSoFar.symbol, autoboxing);
+      if (!isConstructor(method.symbol)) {
+        method.type = typeSubstitutionSolver.applySiteSubstitution(typeSubstitutionSolver.applySiteSubstitution(method.type, superclass), site);
```

Comment:
This should maybe be wrapped into a single call ? (resolvedType, callSite, definition)